src/ast/ast_validations.c

```diff
@@ -1266,51 +1322,57 @@ static void _AST_Pattern_GetDefinedIdentifiers(const cypher_astnode_t *pattern,
 	cypher_astnode_type_t type = cypher_astnode_type(node);
 
 	if(type == CYPHER_AST_RETURN) {
 		/* Only collect aliases (which may be referenced in an ORDER BY)
 		 * from the RETURN clause, rather than all identifiers */
-		_AST_GetReturnAliases(node, identifiers);
+		return _AST_GetReturnAliases(node, identifiers);
 	} else if(type == CYPHER_AST_WITH) {
 		// Get alias if one is provided; otherwise use the expression identifier
-		_AST_GetWithAliases(node, identifiers);
+		return _AST_GetWithAliases(node, identifiers);
 	} else if(type == CYPHER_AST_CALL) {
 		// Get alias if one is provided; otherwise use the expression identifier
-		_AST_GetProcCallAliases(node, identifiers);
+		return _AST_GetProcCallAliases(node, identifiers);
 	} else if(type == CYPHER_AST_MATCH) {
 		/* Collect all identifiers defined by the pattern in the MATCH clause,
 		 * ignoring references in property maps and WHERE predicates. */
 		const cypher_astnode_t *match_pattern = cypher_ast_match_get_pattern(node);
-		_AST_Pattern_GetDefinedIdentifiers(match_pattern, identifiers);
+		return _AST_Pattern_GetDefinedIdentifiers(match_pattern, identifiers);
 	} else if(type == CYPHER_AST_MERGE) {
 		/* Collect all identifiers defined by the path in the MERGE clause,
 		 * ignoring references in property maps and ON CREATE / ON MATCH actions. */
 		const cypher_astnode_t *merge_path = cypher_ast_merge_get_pattern_path(node);
-		_AST_Path_GetDefinedIdentifiers(merge_path, identifiers);
+		return _AST_Path_GetDefinedIdentifiers(merge_path, identifiers);
 	} else if(type == CYPHER_AST_CREATE) {
 		/* Collect all identifiers defined by the pattern in the CREATE clause,
 		 * ignoring references in property maps.  */
 		const cypher_astnode_t *pattern = cypher_ast_create_get_pattern(node);
-		_AST_Pattern_GetDefinedIdentifiers(pattern, identifiers);
+		return _AST_Pattern_GetDefinedIdentifiers(pattern, identifiers);
 	} else if(type == CYPHER_AST_UNWIND) {
 		/* UNWIND only defines its own alias, which is just 'defined' in the query:
 		 * UNWIND [ref_1, ref_2] AS defined RETURN defined */
 		const cypher_astnode_t *unwind_alias_node = cypher_ast_unwind_get_alias(node);
 		const char *unwind_alias = cypher_ast_identifier_get_name(unwind_alias_node);
-		raxInsert(identifiers, (unsigned char *)unwind_alias, strlen(unwind_alias), NULL, NULL);
-	} else if(type == CYPHER_AST_CALL) {
-		_AST_RegisterCallOutputs(node, identifiers);
+		if(!raxInsert(identifiers, (unsigned char *)unwind_alias, strlen(unwind_alias), NULL, NULL)) {
+			ErrorCtx_SetError("Variable `%s` already declared", unwind_alias);
+			return AST_INVALID;
+		}
+		return AST_VALID;
 	} else {
 		uint child_count = cypher_astnode_nchildren(node);
 		for(uint c = 0; c < child_count; c ++) {
 			const cypher_astnode_t *child = cypher_astnode_get_child(node, c);
-			_AST_GetDefinedIdentifiers(child, identifiers);
+			if(_AST_ValidateDefinedIdentifiers(child, identifiers) == AST_INVALID)
+				return AST_INVALID;
```

Comment:
this line not covered by test

Reply:
I don't think it's currently possible to fail at this point, because this `else` condition is for clauses that don't require explicit handling (like `SET`), and it's impossible to nest something that might fail (like `CALL`) into such a case. This still seems like a sensible check to have to me.

Reply:
sow we need ASSERT?